feat: support connecting mcdu to websocket server

tyler58546 opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->
Fixes #1616 
Fixes #6331 

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
* External devices can now send/receive data from the MCDU using a websocket server.
* Added server that can be used to remotely control the MCDU from external devices and print to real printers

## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->

![image](https://user-images.githubusercontent.com/16512508/146491472-51e03380-6482-4657-bae0-7209e8a3d89f.png)

## References
<!-- You should be making changes based on some kind reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->


## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->
1. Start a flight
2. Open your community folder and navigate to `flybywire-aircraft-a320-neo\MCDU SERVER\` and open `server.exe`
3. Follow the instructions in the window that opens

Things to check:
* You are able to control the MCDU from another device on your network (the IP address given by the server is correct)
* (If you have both ethernet and wifi) You are able to connect from another device on your network when using both connections
* All MCDU pages are formatted correctly and function as intended on both the real MCDU and the external device
* Please test the MCDU client on a variety of devices and browsers

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

ninoz commented:
I merged this into my local copy of FBW and compiled and installed the app on my iPad.  It's such a good addition the the system, I can now fiddle with the MCDU without having to change my camera.  The build.sh has changed so that needs looked at and the selection buttons at the side of the screen don't seem to register but other than that its fantastic!

Epic work on this @tyler58546

ninoz commented:
I've tested this again and it all works as expected :)  

I did have to add  "es5lib": "^0" into package.json to build, sorry im not sure how to edit the PR to add that in

Hadi-1624 commented:
Is it still possible to test this? I would appreciate any instructions to do so as I am very interested in this feature.

wpine215 commented:
Any update on this PR?

frankkopp commented:
Started doing documentation for his.

Some feedback:

1. Awesome 🙂
2. maybe center the MCDU in the HTML - see my iPad screenshot below
3. the IP address the console gives did work for me as I do not use 192.168..... is it possible to better detect the local address? 
4. how to configure the firewall will be part of my documentation

![image](https://user-images.githubusercontent.com/16833201/145419070-b6656a73-2b27-4b86-82ed-43582d531ed0.png)

tyler58546 commented:
* You can now view only the MCDU screen by tapping the title line
* Fixed centering

RichardPilbery commented:
I know it's not ready for QA review, but I took this PR for a spin last night. Full flight, no issues. Really responsive throughout. 

Checked the following pages (and any sub-pages under):
+ MCDU MENU
	+ ATSU
	+ AOC Menu
		+ INIT/PRES
		+ FREE TEXT
		+ WEATHER REQUEST
		+ ATIS
		+ W/B
		+ FREE TEXT
		+ RECEIVED MESSAGES
		+ SENT MESSAGES

	+ DATALINK STATUS
	+ COMM MENU

	+ ATC
		+ Emergency
		+ Reports
		+ MSG Record
		+ Connection
		+ REQUEST
	+ AIDS
	+ CFDS
		+ AVIONICS STATUS
		+ SYSTEM REPORT / TEST
+ DATA
	+ POSITION MONITOR
	+ IRS MONITOR
	+ GPS MONITOR
	+ A/C STATUS
	+ CLOSEST AIRPORTS
	+ WAYPOINTS
	+ NAVAID
	+ STORED WAYPOINTS

All work fine, just some visual issues:
## Closest aiports page
![MCDU-6](https://user-images.githubusercontent.com/5834254/145611250-5b9f882c-3bf8-4798-974c-a77abb5f9286.png)
![MCDU-5](https://user-images.githubusercontent.com/5834254/145611306-8ee42ab5-e35c-410b-b449-6503cacc0734.png)

## Init review
Apart from the visual artifact komp pointed out, the 1/2 indication of multiple pages is missing top right.
![MCDU-7](https://user-images.githubusercontent.com/5834254/145611399-ad6f0fdb-47fd-4e88-a7bd-11fb77160105.png)
![MCDU-8](https://user-images.githubusercontent.com/5834254/145611395-f98ddc73-462d-46e3-832c-e898c2b6fe5a.png)

## AOC MSG DISPLAY
ATIS is a bit cramped
![MCDU-2](https://user-images.githubusercontent.com/5834254/145612869-d1485fea-ad7e-4ab3-b87f-f958e20009e6.png)


## Flightplan page
Middle pink asterix are running into the times
![MCDU-3](https://user-images.githubusercontent.com/5834254/145611787-f1cfbfd6-5cfe-4077-8b6a-90ceae451576.png)

## iPhone 12
![iphone-12-add-2-homescreen](https://user-images.githubusercontent.com/5834254/145612134-b3c5b1b1-f70f-4a85-a65c-27a023423fc1.png)
![iphone-12-chrome-crowded-pink-star](https://user-images.githubusercontent.com/5834254/145612152-229dd6c5-10ec-4da5-ae6a-4cd29e6364ac.png)
![iphone-12-safari](https://user-images.githubusercontent.com/5834254/145612171-1341213b-bea8-4433-8aa2-1134170511ec.png)

## 2020 iPad 10''
MCDU is cropped unless 'Add to HomeScreen option used'
Small visual defect on second image.
Might be good to resize based on screen height and centre the MCDU on the page.
![ipad-mcdu-not-fit](https://user-images.githubusercontent.com/5834254/145612480-0cee574c-ed2d-4faa-9f99-7eb829eb3efc.png)
![ipad-take-off-rwy-visual-error](https://user-images.githubusercontent.com/5834254/145612523-116f54f4-e11a-4869-890f-ca5b7a9443e1.png)

RichardPilbery commented:
Last thing: Network IP address is incorrect. In this instance, it should be 192.168.1.210:8125
![MCDU-1](https://user-images.githubusercontent.com/5834254/145613215-4ed22bd1-161e-4e69-a034-3c7280bcaa09.png)

tracernz commented:
> * You can now view only the MCDU screen by tapping the title line

Is it possible to make a query string param that asks for only the MCDU display? That would be appreciated by cockpit builders.

tyler58546 commented:
* Added dedicated fullscreen url (`x.x.x.x:8125/fullscreen`)
* Fixed left title and page number
* Fixed MCDU getting cut off at the bottom on iPad
* Fixed other layout issues

FinniganH commented:
This appears to not want to load fully in samsung browser. Samsung browser loads the MCDU, but the screen is blank. Open it in chrome and it works just fine.

Samsung Browser wasn't cofnigured one way or another on privacy, so it shouldn't be overly restrictive.

tracernz commented on .gitignore:
```diff
@@ -53,3 +53,5 @@ node_modules
 .env
 src/instruments/buildSrc/custom/*
+/flybywire-aircraft-a320-neo/MCDU SERVER/
+/src/mcdu-server/client/build
```
Missing trailing newline

tracernz commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_AvailableArrivalsPage.js:
```diff
@@ -271,6 +256,21 @@ class CDUAvailableArrivalsPage {
+                ["{sp}APPR", "STAR{sp}", "{sp}VIA"],
+                [selectedApproachCell + "[color]" + selectedApproachCellColor, selectedStarCell + "[color]" + selectedStarCellColor, "{sp}" + selectedViasCell + "[color]" + selectedViasCellColor],
+                [viasPageLabel, "TRANS{sp}"],
+                [viasPageLine, selectedTransitionCell + "[color]" + selectedTransitionCellColor],
```
Did you mean to remove the old `[color]` everywhere? I see some are and some aren't e.g. here. I haven't noted the rest that remain.

tyler58546 replied:
No, I only removed it in places where it was causing issues.

tracernz commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js:
```diff
@@ -1429,5 +1457,106 @@ class A320_Neo_CDU_MainDisplay extends FMCMainDisplay {
+            clearTimeout(this.socketTimeout);
+        }
+
+        const url = "127.0.0.1:8080";
```
Do you want me to add a commit with a flyPad setting for this? Maybe also a toggle to enable/disable?

tyler58546 replied:
I already made one previously (see earlier commits) but ended up taking it out because I wanted to make it work without having to configure anything in the sim (just run server.exe and open the url). The only issue with this is the user might want to change the port in the event that it is taken, so I am considering adding it back.

tracernz replied:
Maybe we could make it default to `http://localhost:8080` and enabled... and power users can choose a different URL in the flypad if there is a port clash?

tyler58546 replied:
Alright, I added an EFB option for the web socket url and a command line argument for the the server port.

tracernz commented on src/mcdu-server/client/src/McduScreen.css:
```diff
@@ -0,0 +1,171 @@
+    --displayGreen: #00ff00;
+    --displayMagenta: #ff94ff;
+    --displayRed: #ff0000;
+    --displayYellow: #ffff00;
```
Is it too cheeky to include these defs from `src\instruments\src\Common`?

tyler58546 replied:
I'm not sure that's the best idea. If someone makes changes to that file it's unlikely they will test this too.

tracernz commented on src/mcdu-server/index.js:
```diff
@@ -0,0 +1,251 @@
+path.join(__dirname, '../../node_modules/pdfkit/js/data/Helvetica.afm');
+
+const sumatraPdfPath = path.join(os.tmpdir(), 'SumatraPDF.exe');
+fs.copyFileSync(path.join(__dirname, '../../node_modules/pdf-to-printer/dist/SumatraPDF.exe'), sumatraPdfPath);
```
Can you explain this part a little?

tyler58546 replied:
The library that I'm using to to print, _pdf-to-printer_ uses an open source program called _SumatraPDF_ to communicate with the printer driver. This code extracts `SumatraPDF.exe` to a temporary folder so it can be run by _pdf-to-printer_.

tracernz replied:
So it can't run it from the node_modules directory?

tyler58546 replied:
It can't run from _pkg_'s virtual filesystem when it's packaged in an exe. Windows needs it to be its own file.

tracernz replied:
Ah, that makes sense!

tracernz reviewed (commented):
Can we use the fonts in `src\fonts`?

tyler58546 commented:
> Can we use the fonts in `src\fonts`?

Done. Now it copies from `html_ui/fonts` during build.

tracernz commented on src/instruments/src/EFB/Settings/Settings.tsx:
```diff
@@ -398,6 +400,21 @@ const SimOptionsPage = () => {
+                            value={mcduServerUrl}
+                            noLabel
+                            onChange={(event) => {
+                                setMcduServerUrl(event);
```
Should we do some basic validation with https://developer.mozilla.org/en-US/docs/Web/API/URL ?

tyler58546 replied:
`SimpleInput` does not have a good system for this type of input validation and I think creating one would be outside the scope of this PR.

frankkopp commented:
Can I suggest to make the websocket port not part of the EFB settings? This is really advanced and should not be mixed with the other settings in this way. Either have another tab for this setting or use config file (similar to autopilot).

I can almost guarantee that users will mistake this setting for the http browser URL to be used for the MCDU Web Interface.

k0deine commented:
> This appears to not want to load fully in samsung browser. Samsung browser loads the MCDU, but the screen is blank. Open it in chrome and it works just fine.
> 
> Samsung Browser wasn't cofnigured one way or another on privacy, so it shouldn't be overly restrictive.

Also the case for older devices such as IOS 9.3, screen stays blank.
Any older browser such as Chrome on the same devices does not render it well either
You can solve it by running the browser to the MCDU on a desktop device and using remote screen software on the mobile device..

frankkopp commented:
Updated documentation draft: 
https://docs-git-fork-frankkopp-feat-web-mcdu-flybywire.vercel.app/fbw-a32nx/feature-guides/web-mcdu/

frankkopp commented:
As it will not be possible to test and support all browsers there will be a list added to the documentation page and a general statement that it needs modern standard browsers. 

@tyler58546 please add this to the testing instruction that it should be tested with as many browsers as possible and they should be listed in the reports so we can build a list of compatible browsers.

tracernz commented:
> Can I suggest to make the websocket port not part of the EFB settings? This is really advanced and should not be mixed with the other settings in this way. Either have another tab for this setting or use config file (similar to autopilot).
> 
> I can almost guarantee that users will mistake this setting for the http browser URL to be used for the MCDU Web Interface.

There is no way for the user to resolve a port conflict without it.. i.e. they cannot use this at all if something else is listening on 8080 (not uncommon).

frankkopp commented:
> > Can I suggest to make the websocket port not part of the EFB settings? This is really advanced and should not be mixed with the other settings in this way. Either have another tab for this setting or use config file (similar to autopilot).
> 
> > 
> 
> > I can almost guarantee that users will mistake this setting for the http browser URL to be used for the MCDU Web Interface.
> 
> 
> 
> There is no way for the user to resolve a port conflict without it.. i.e. they cannot use this at all if something else is listening on 8080 (not uncommon), unless we provide the setting.

We do not have to use such a common port as 8080 :)

At least we should rename the setting and maybe add a note that this is not the port for the browser. I tried to make this very clear in my draft docs page but we know users often don't read docs

tyler58546 commented:
Maybe we could change the setting to just be the port instead of the full URL. What do you think @tracernz?

tracernz commented:
> Maybe we could change the setting to just be the port instead of the full URL. What do you think @tracernz?

Yep, works for me, and solves the outstanding issue above. I have reviewed all the code so once you've done this I'll approve and it's ready to test. 👍

omrygin commented:
I don't know if it is implemented or intended to be, but home cockpit builders might benefit from an option to send only the display over a websocket. Just something to note for the future.

tracernz commented:
> I don't know if it is implemented or intended to be, but home cockpit builders may benefit from an option to send only the display over a websocket. Just something to note for the future.

It is, add `fullscreen` to the URL

aguther commented:
I‘m using this feature for a while now and from my point it‘s ready to go, had no issues found.

SVitalini commented:
Using the latest experimental version from today, whenever i open the server.exe i get this error:
[https://imgur.com/a/GByttQ1](url)

SVitalini commented:
> Using the latest experimental version from today, whenever i open the server.exe i get this error: [https://imgur.com/a/GByttQ1](url)
> 
> EDIT: Printspool service was deactivated as i have no printer installed. Restartet the service and it works now!

FIXED by enabling Printspool Windows service! Maybe worth a note in the documentation as others might run in this issue aswell?

mehoygg commented:
Quality Assurance Tester/Trainee Report

Discord          : mehoygg#1242
Object of testing: #3467
Tier of Testing  : 1
Date             : 17/12/21

Testing Process: 
Spawned in C/D at KLAX. 
I went to the MCDU Server folder in the A/C folder in Community.
I launched it and got it running.
I used both an iPad Mini and an iPhone 13 Pro with Safari and Chrome.
As soon as the web browser MCDU pops up, its black, signaling no power.
I then proceeded to give the A/C EXT Power and both the MCDU in sim and on the devices turned on. 
I was able to set up a flight plan from KLAX/KSFO all via the web browser.  Really neat stuff!  I set up the whole flight via the browser.  Init/Init B, loading PAX, setting up the PERF Page, all of it. Really convenient to be able to set up the MCDU and not have to be looking down at it in the sim.  
I also was able to go onto the web browser with a second device and the information was still there as I had inputted it and was able to change values etc.

Same IP address is given and works for my devices.  Great work and its a really cool feature.

Negatives:       
None besides my big fingers mis-clicking a few keys on the devices.  Nothing a CLR couldnt fix. 

Testing Results: 
Passed

Conclusions: 
Had an issue with the iPhone but was on my end.  Once I noticed it, I did rectify it and retested.  Works.


Media:       
Posted on Discord shots from ipad/iphone

jcalor commented:
Quality Assurance Tester Report

Discord          : Calo#6816
Object of testing: #3467 
Tier of Testing  : 1
Date             : 17/12/2021

Testing Process: 

- Opened server.exe as administator
- Followed its steps and tried on Chrome, Mozilla Firefox, Microsoft Edge and on my phone (Android).
- Been interacting with MCDU, importing some Simbrief FPs, trying to "broke" MCDU. Works!

Negatives: server.exe link to find your local IP Address does not work

![image](https://user-images.githubusercontent.com/61759243/146615078-4254dfaa-8088-484e-9436-139894deab52.png)

Testing Results: Passed

Conclusions: Everything works except that small link, but is quite simple to find your IP Address: open command line - ipconfig and there you go :)

FlightSimmer7700 commented:
Quality Assurance Tester

Discord: FlightSimmer7700#5438
Object of testing: #3467 
Tier of Testing: 1
Date: 17/12/21

Testing Process:
Loaded in, opened server.exe.
Pc (same device)
Chrome- worked
Firefox-worked
microsoft edge- worked

Samsung phone:
Google chrome- worked
Firefox- worked
duckduckgo-worked
Samsung internet-does not work

Ipad:
Safari-worked
Chrome- worked
Edge-worked
bing-worked
firefox- worked
duckduckgo-worked
brave-worked

Negatives: 
- The mcdu display is brighter than usual when off and on
- Samsung internet does not work
- Weird issues Ive had on my ipad but I believe they are the IOS issue as these do not occur when Im on my phone
- IRS INIT went blank for some reason

Testing Results: Not Passed

Conclusions: Its an amazing feature thank you so much, I had tons of fun using it but it has some weird issues sometimes out of nowhere
Media:

MCDU screen brightness:
![image](https://user-images.githubusercontent.com/70103461/146620636-c8ef6417-34eb-403c-9ae9-2aa213fa1fa5.png)

IRS INIT in sim:
![image](https://user-images.githubusercontent.com/70103461/146620683-f1e9cd53-9020-492f-8049-4837bd8dacc6.png)
IRS INIT on ipad:
![image](https://user-images.githubusercontent.com/70103461/146620768-893a8704-5a72-4ff2-87f0-9b3bcc65c533.png)

When entering something into a field on the ipad, doesn't matter what the browser was it keeps this part like that and is only fixed when I reload the page
![image](https://user-images.githubusercontent.com/70103461/146620822-0c4021a7-be03-470b-aadd-7ac790c37adb.png)

On the other hand, similarly when deleing the stuff out of the field it leaves this weird part cut off
![image](https://user-images.githubusercontent.com/70103461/146620850-004e0873-2618-40c3-ac3c-0d92e1730fce.png)

^Nothing removed but same effect when i entered the ils frequency 
Should look like: 
![image](https://user-images.githubusercontent.com/70103461/146620897-d0770fce-6a81-4e7b-b51f-c7fc74e1f019.png)
What it looks like:
![image](https://user-images.githubusercontent.com/70103461/146620910-0703a944-dd93-476c-b517-185ebbc8ee89.png)

Samsung internet while the plane is on and it works on every other browser
![image](https://user-images.githubusercontent.com/70103461/146621008-f1f3b90b-070e-4697-b8c5-208e8287702d.png)

frankkopp commented:
> Testing Process:
Loaded in, opened server.exe.
Pc (same device)
Chrome- worked
Firefox-worked
microsoft edge- worked

> Samsung phone:
Google chrome- worked
Firefox- worked
duckduckgo-worked
Samsung internet-does not work

> Ipad:
Safari-worked
Chrome- worked
Edge-worked
bing-worked
firefox- worked
duckduckgo-worked
brave-worked

What is the Bing browser? Bing is a search engine? Couldn't find a brwoser with the name in the App Store.
What is the Samsung Internet browser called? Just "Internet"?

frankkopp commented:
Anybody able to test with Linux browsers?

FlightSimmer7700 commented:
> > Testing Process:
> > Loaded in, opened server.exe.
> > Pc (same device)
> > Chrome- worked
> > Firefox-worked
> > microsoft edge- worked
> 
> > Samsung phone:
> > Google chrome- worked
> > Firefox- worked
> > duckduckgo-worked
> > Samsung internet-does not work
> 
> > Ipad:
> > Safari-worked
> > Chrome- worked
> > Edge-worked
> > bing-worked
> > firefox- worked
> > duckduckgo-worked
> > brave-worked
> 
> What is the Bing browser? Bing is a search engine? Couldn't find a brwoser with the name in the App Store. What is the Samsung Internet browser called? Just "Internet"?

Well I went into the app store and was installing chrome, firefox, brave and so on and bing came up so I installed and tried that too

gustavocontin commented:
> Anybody able to test with Linux browsers?

I tested it on my raspberry, Chromium browser and Raspbian 10 OS (old versions). 
It worked perfectly without any problem. 

Congratulations on the work of the Flybywire team.

ssewell commented:
It would be useful to cockpit builders to be able to explicitly specify the width and height when using fullscreen mode to address the various real-world LCD screens. Currently, the calculated aspect ratio of the DIV in which the content is rendered with result in black bars if the virtual screen doesn't exactly match the physical screen upon which it's been drawn.

frankkopp commented:
@tyler58546 could you please rebase or merge from master. EFB Settings page for Sim Options has changed and been split into two page. It should move under Dynamic Registration Decal in Sim Options.
It is necessary for me to be able to take screenshots for docs.

tyler58546 commented:
> Negatives:
>
> - The mcdu display is brighter than usual when off and on

I was not able to reproduce this. It looks like the airport lights were just shining on it.
> - Samsung internet does not work

Since I don't have a device with this browser, I am not able to fix this.
> - Weird issues Ive had on my ipad but I believe they are the IOS issue as these do not occur when Im on my phone

Should be fixed now.
> - IRS INIT went blank for some reason

I was not able to reproduce this issue, but I was able to fix the issue with the IRS INIT page not refreshing that you showed me on Discord.

FlightSimmer7700 commented:
Quality Assurance Tester

Discord: FlightSimmer7700#5438
Object of testing: #3467 
Tier of Testing: 1
Date: 25/12/21

Testing Process: 
Completed a whole flight while trying both on using my phone, ipad and through in sim
(have NOT tested with real printer as I don't have one)
Negatives: None that are very important but I feel the IRS page is still quite late with responses, maybe you can take a look at that again as I feel its still quite slow
Testing Results: Passed

Conclusions: LGTM works great, very good work!

Azores77 commented:
Can I try it or this is just for testers? :)

gustavocontin commented:
> Can I try it or this is just for testers? :)

I'm not a tester but I can test this amazing feature, just check the instructions at the beginning of this page.

Azores77 commented:
> page.

I did but I don’t see anything to download.
I’m not familiar with this GitHub at all.
I think I’ll wait until this feature is available.

DavidEGTE commented:
@Azores77 the feature is available via installer on dev version, just need to update :)

c0nnex commented:
Nice Approach, however it's pretty much limited becasue it relies that only the included server.exe will react to the data.

I'll move my SPAD.neXt implementation of the FBW Cdu for Hardware-implemention to this websocket approach, however, since SPAD.neXt is exposing a global webinterface and I do not want to expose an extra interface "just" for the fbw.

So I would suggest to either 
* change the webSocket url to something like "ws://127.0.0.1:PORT/FBW_CDU" (maybe include aircraft in it? FBW_A32NX_CDU ?)
* or let the user configure a webSocket-Subpath, just like they can configure the port.

Nota Bene: I don't think 8080 is a good choice as default port. It's widely and commonly used for many many web-interfaces. This might  bring up a lot of support "Web CDU not working" at your side.

frankkopp commented:
Hi Ulrich,

great feedback. Will follow up with the team and get back to you. 

Cheers Cdr_Maverick

tracernz commented:
The web socket port is a setting in the flyPad. Originally it was going to be a URL but it was felt this would be too confusing for some users.
